src/pass/arg_binder.cc

```diff
@@ -44,36 +49,40 @@ bool ArgBinder::Bind_(const Expr& arg,
       } else {
         (*def_map_)[v] = value;
       }
-      return true;
+      (*new_variable_added) = true;
+      result = true;
     } else {
-      BinderAddAssert(it->second == value, arg_name, &asserts_);
+      result = BinderAddAssert(it->second == value, arg_name);
     }
   } else {
-    BinderAddAssert(arg == value, arg_name, &asserts_);
+    result = BinderAddAssert(arg == value, arg_name);
   }
-  return false;
+  (*new_variable_added) = false;
+  return result;
 }
 
-void ArgBinder::Bind(const Expr& arg,
+bool ArgBinder::Bind(const Expr& arg,
                      const Expr& value,
                      const std::string& arg_name,
                      bool with_let) {
-  Bind_(arg, value, arg_name, with_let);
+  bool new_var_added = false;
+  return Bind_(arg, value, arg_name, with_let, &new_var_added);
 }
 
-void ArgBinder::BindArray(const Array<Expr>& arg,
+bool ArgBinder::BindArray(const Array<Expr>& arg,
                           const Array<Expr>& value,
                           const std::string& arg_name) {
   CHECK_EQ(arg.size(), value.size())
       << "Argument " << arg_name << " array size mismatch";
   for (size_t i = 0; i < arg.size(); ++i) {
     std::ostringstream os;
     os << arg_name << "[" << i << "]";
-    this->Bind(arg[i], value[i], os.str());
+    COND_CHECK(quit_on_assert_, this->Bind(arg[i], value[i], os.str()));
```

Comment:
can't we just do CHECK(!quite_on_assert_ && condition)?

Reply:
No. Implicit in COND_CHECK is that if (!quit_on_assert_ && !condition) then we return false to inform the caller what happened.

Reply:
I see, this behavior is somewhat hard to understand (return hides in the macro itself).  I  understand why it is favorable here.  Maybe we should make it explicit(three lines vs one line)

```c++

bool MyChecker() {
    if (condition_false) {
       if (quite_on_assert) return false;
       LOG(FATAL) << "msg"; 
    }
}
```

Reply:
just to favor readability over the conciseness

Reply:
Readability is definitely important. In this case, replacing a one-liner with a four-liner makes more of the code explicit, and therefore, more readable in one aspect. However, that replacement will increase the code size by a few hundred repetitive lines at least (there are tens of these COND_CHECKs in tensorize.cc). It will make the main logic of the code be overshadowed by a large number of if statements. And that makes the code less readable in another aspect. Overall, IMHO, using COND_CHECK will make the code more readable than replacing by four-liner.
One compromise we can make is to always pass "return false" explicitly to COND_CHECK (note that sometimes we need to pass "continue" or "break" in tensorize.cc as that parameter)

Note COND_CHECK is not without precedent. CHECK macros also have a fair amount of implicit code in them. Admittedly, COND_CHECK has slightly more implicit code in them. But I believe after a developer looks up the macro definition a couple of time, they can read the code more naturally.

Reply:
Considering my opinions above, please let me know what's acceptable to you and I'll adopt it. I don't want  the discussion to slow us down. Thanks!

Reply:
OK, I think we can do the macro as long as we add proper doxygen comment  to properly document what is the usage example, as well as proper comment on the binder.

```c++
\code
\endcode
```

Now, I have less idea on how to solve the linter issue. Maybe ```/**/``` style comment will help